1388: added Azure AD authentication

avas opened the pull request:
The authentication is made using OpenID Connect. 

**NOTE**: Some additional set-up is required to test this feature - please see the [walkthrough](https://github.com/VirtoCommerce/vc-platform/issues/1388#issuecomment-423525628) in the issue comment for further details.

Changes made for #1388 to add the external authentication with Azure AD account:
* Added some Azure AD-related properties to the Web.config and corresponding properties in `AuthenticationOptions` class. Also added code to map it together to the `Startup` class.
* Added `Microsoft.Owin.Security.OpenIdConnect` to `VirtoCommerce.Platform.Web`. It is initialized in `OwinConfig.cs`.
* Added the `ExternalLoginController` that redirects the user to the Azure AD sign-in page and creates/updates the VC platform account if needed. It uses the `ISecurityService` to work with platform accounts. All dependencies are injected with DI - I enabled it for MVC in `VirtoCommercePlatformWebBootstrapper.cs`.
* Login dialog: added link for Azure AD sign-in. It points to the `ExternalLoginController.SignIn` action.

tatarincev commented on VirtoCommerce.Platform.Web/VirtoCommercePlatformWebBootstrapper.cs:
```diff
@@ -49,7 +51,7 @@ public override void Run(bool runWithDefaultConfiguration)
 
             //registering Unity for MVC
-            //DependencyResolver.SetResolver(new UnityDependencyResolver(container));
+            DependencyResolver.SetResolver(new MvcUnityDependencyResolver(Container));
```
Is this change necessary?

avas replied:
Yes, it enables DI for regular (non-WebAPI) MVC controllers. It is used to inject dependencies to the `ExternalLoginController`.

tatarincev commented on VirtoCommerce.Platform.Web/VirtoCommerce.Platform.Web.csproj:
```diff
@@ -1189,6 +1204,7 @@
     <Folder Include="App_Data\Modules\" />
     <Folder Include="Filters\" />
     <Folder Include="Modules\" />
+    <Folder Include="Views\ExternalLogin\" />
```
Empty folder

tatarincev commented on VirtoCommerce.Platform.Web/Scripts/app/security/login/login.tpl.html:
```diff
@@ -38,6 +38,11 @@
                     {{authError}}
                 </div>
             </div>
+            <div class="form-group">
```
I think we shouldn't display this action if  `VirtoCommerce:Authentication:AzureAD.Enabled` setting is off

tatarincev commented on VirtoCommerce.Platform.Web/Controllers/ExternalLoginController.cs:
```diff
@@ -0,0 +1,200 @@
+        {
+            var externalLogins = platformUser.Logins?.ToList() ?? new List<ApplicationUserLogin>();
+
+            var externalLogin = ConvertExternalLoginInfoToExternalLogin(externalLoginInfo);
```
Are you sure what this logic won't create duplicate records after each external login?

avas replied:
Well, it shouldn't - if the user already had an external login entry, they will sign in successfully (`externalLoginResult == SignInStatus.Success` on line 83), and this method won't get called. And even if it *will* create duplicate record and attempt to save it, it will lead to duplicate record in `dbo.AspNetUserLogins` table - such record will violate the primary key constraint, and RDBMS will not allow to save it.
![image](https://user-images.githubusercontent.com/1835759/45994632-b6eb6700-c0be-11e8-9e5b-88c21643247b.png)

avas replied:
Reworked the code that adds external logins. Now there won't be any external login duplications.

tatarincev commented on VirtoCommerce.Platform.Web/Controllers/ExternalLoginController.cs:
```diff
@@ -0,0 +1,200 @@
+            }
+
+            var signInManager = _signInManagerFactory();
+            var externalLoginResult = await signInManager.ExternalSignInAsync(externalLoginInfo, false);
```
There is some code duplications in `SignInExternalUser` and `CreateOrUpdatePlatformUser` methods
```
 await signInManager.SignInAsync(aspnetUser, true, true);
 await _eventPublisher.Publish(new UserLoginEvent(platformUser));
```

 I think it can be eliminated after some  little code refactoring to the next logic flow
1.  `ExternalSignInAsync` 
2. If exist user try to add external login and update user 
3. else need to create new user with external logins
4. user = userManager.FindByLoginInfo(loginInfo)
5. signInManager.SignIn(user)
6. publish event

avas commented:
* Refactored `ExternalLoginController` a bit to reduce code duplication.
* Reworked the code to display external login link(s) on login dialog (see 87c7148):
    * Added `SecurityController.GetExternalLoginProviders` method (available at `/externalloginproviders`), so that frontend could request available external providers;
    * Login dialog now requests external login providers and displays `Sign in with (provider name)` link for each provider. This way Azure AD sign-in link won't be displayed on login dialog when Azure AD authentication is disabled. It will also make it easier to add more external sign-in providers if needed.

avas commented:
@tatarincev Of course, I'll do it.

avas commented:
@tatarincev Well, the code looks much cleaner and mostly works. But it fails in one particular case: if the VC platform account with Azure AD login is already created (e.g. manually by admin), and that user attempts to log in externally. In this case platform will attempt to create a new user account and will fail, because that account already exists - it is just missing the `ExternalLogin`.
![image](https://user-images.githubusercontent.com/1835759/46357294-be96a700-c68e-11e8-84c3-bc2b87c0cb42.png)

Will this case be used by end users of this feature? If so, contents of the `if (externalLoginResult == SignInStatus.Failure) { ... }` block will need one more `if`.

avas commented:
I've made some changes to fix this error. The code works fine for me, but looks a bit uglier.
@tatarincev Can you please check changes made in 0115847?

avas commented:
@tatarincev Turns out that `User.Identity.Name` can be empty after external sign-in, even if that sign-in was successful. So I guess we'll really have to use `userName` extracted from user claims.
![image](https://user-images.githubusercontent.com/1835759/46363909-86976000-c69e-11e8-96d3-c8e448b693a7.png)

I also wasn't sure whether the `ApplicationUserExtended.Logins` can be null, but turns out it can't - at least I didn't achieve that. So the code is okay.

I've pushed my changes in 9912229. Can you please take a look?

tatarincev commented on VirtoCommerce.Platform.Web/Controllers/Api/SecurityController.cs:
```diff
@@ -588,6 +592,24 @@ public async Task<IHttpActionResult> UnlockUserAsync(string id)
             return Content(result.Succeeded ? HttpStatusCode.OK : HttpStatusCode.BadRequest, result);
         }
 
+        [HttpGet]
```
I think this method  can be moved into ExternalLoginController
